src/diffusers/pipelines/pipeline_loading_utils.py

```diff
@@ -947,3 +948,18 @@ def _get_ignore_patterns(
             )
 
     return ignore_patterns
+
+
+def model_has_device_map(model):
```

Comment:
This is the most important bit to take note of. And I have added tests for it too (`test_sharded_components_can_be_device_placed()`).